### DISCUSSION TITLE
Mark field as covariant / read-only to allow Liskov substutution

### DISCUSSION THREAD

**Discussion by fritzo** [pydantic/pydantic#5870] [Feature Request]
I'd like to create a morally read-only model with a field that's a list of strings.  And I'd like to follow [Postel's law](https://en.wikipedia.org/wiki/Robustness_principle) and initialize my `list[str]` field from any `Sequence[str]`, which seems like very Pydantic-idomatic behavior:
```python
from collections.abc import Sequence

class MyModel(BaseModel):
    strings: list[str]
    
    @validator
    def check_strings(cls, strings: Sequence[str]) -> list[str]:
        return list(map(str, strings))
```
However when I initialize `MyModel(my_list)` where `my_list: MyList[MyStr]` and `MyList` is a subclass of `list` and `MyStr` is a subclass of `str`, mypy complains about list being invariant rather than covariant:
```
error: Argument "strings" to "MyModel" has incompatible type "MyList[MyStr]"; expected "List[str]"  [arg-type]
```
I tried to replace strings with a covariant type hint
```python
class MyModel2(BaseModel):
    strings: Sequence[str]
```
but then mypy complains about downstream consumers of `MyModel2.strings` who expect a list, i.e. mypy is unware of my Postel-style casting to a safe type.

Is there a way I can mark a field as read-only / covariant and get mypy to understand?

**Comment by dmontagu** [comment]
I think this isn't a pydantic/mypy/python shortcoming so much as it is a fundamental type safety issue.

What I would do to be compliant with the type system is keep the annotation as `Sequence[str]` and use `m_strings: list[str] = list(m.strings)` or similar (where `m` is an instance of `MyModel2`). You have to do this for a good reason — lists are mutable, and if you were to call `append` with a `str` on your `MyList[MyStr]`, you'd end up with something that is no longer a `MyList[MyStr]`.

If you don't want to pay the price of copying the list and know it won't be used in a type-unsafe way, you can use `m_strings = typing.cast(list[str], m.strings)` or even just `# type: ignore`, but obviously these leave you open to uncaught type errors as you make changes to the code in the future.
